Fix crash on deriveaddresses when index is 2147483647 (2^31-1)

muxator opened the pull request:
This PR is a proposal for fixing #26274 (better described there).

The problem is due to a signed int wrapping when the `index` parameter of the `deriveaddresses` RPC call has the value `2^31-1`.

```C++
for (int i = range_begin; i <= range_end; ++i) {
```

* the first commit adds a "temporary" test case (`test/functional/rpc_deriveaddresses_crash.py`) that shows the crash, and can be used to generate a core dump;
* the second commit fixes the problem giving an explicit size to the `i` variable in a for loop, from `int` to `int64_t`. The same commit also removes the ephemeral test case and adds a passing test to `test/functional/rpc_deriveaddresses.py`, in order to prevent future regressions.

This is my first submission to this project and I do not know its conventions. Please advise if something needs to be changed.

Commit:
rpc: add non-regression test about deriveaddresses crash when index is 2147483647

This test would cause a crash in bitcoind (see #26274) if the fix given in the
previous commit was not applied.

instagibbs commented:
Initial NACK on approach, as I believe that's not BIP32 compatible?

e.g. https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki#private-parent-key--private-child-key

"ser32(i)" is implicitly stating that `i` must be 32 bits long maximum. Throwing an RPC error on detection of potential overflow seems best.

instagibbs commented:
@achow101 should descriptors barf when asked to `Expand` something invalid?

MarcoFalke commented:
I don't think this changes behaviour. It only fixes a sanitizer warning

achow101 commented:
Concept ACK

Your first commit introduces a test file that ends up being removed in the second commit. I think it is better to just have the fix in a first commit, and the test in a second. Reviewers can then cherry-pick the test commit onto master to verify that it fails without the first commit. We generally don't want to be introducing stuff in a commit that is going to be removed immediately afterwards. It makes searching through history harder.

Commit:
rpc: fix crash in deriveaddresses when derivation index is 2147483647

2147483647 is the maximum positive value of a signed int32, and - currently -
the maximum value that the deriveaddresses bitcoin RPC call accepts as
derivation index due to its input validation routines.

Before this change, when the derivation index (and thus range_end) reached
std::numeric_limits<int_32_t>::max(), the "i" variable in the for cycle (which
is declared as int, and as such 32 bits in size on most platforms) would be
incremented at the end of the first iteration and then warp back to
-2147483648. This caused SIGABRT in bitcoind and a core dump.

This change assigns "i" an explicit size of 64 bits on every platform,
sidestepping the problem.

Fixes #26274.

muxator commented:
> I think it is better to just have the fix in a first commit, and the test in a second.

@achow101, I've rewritten the history organizing it as you proposed, thanks for the advice.

yancyribbens commented:
Concept ACK

`i` should be the same size data type as [range begin](https://github.com/bitcoin/bitcoin/blob/master/src/rpc/output_script.cpp#L252) and [range end](https://github.com/bitcoin/bitcoin/blob/master/src/rpc/output_script.cpp#L253) to prevent an overflow.

willcl-ark commented:
tACK both commits cherry-picked onto v24.0rc2.

I was able to reproduce the crash before the patch with `./src/bitcoin-cli deriveaddresses "wpkh([d34db33f/84h/0h/0h]xpub6DJ2dNUysrn5Vt36jH2KLBT2i1auw1tTSSomg8PhqNiUtx8QX2SvC9nrHu81fT41fvDUnhMjEzQgXnQjKEu3oaqMSzhSrHMxyyoEAmUHQbY/0/*)#cjjspncu" "[2147483647, 2147483647]"`.

Afterwards this call succeeded deriving address `bc1qpmmpntm7hjmfulr8tuqtyzv9s5ahn9lk857qy4`.

achow101 commented:
ACK 9153ff3e274953ea0d92d53ddab4c72deeace1b1

MarcoFalke commented:
Marked for backport, so that it is easier to run previous branches with sanitizers enabled

MarcoFalke commented:
Actually it is unclear whether this is a behaviour change. It is UB and my compilers happened to optimize the bug away, but there is no guarantee that other compilers or compiler options will do so as well.

fanquake commented:
Backported to 24.x in #26410.

fanquake commented:
Backported to 23.x in #26411.

fanquake commented:
Backported to 22.x in #26413.
